src/coreclr/inc/corhlprpriv.h

```diff
@@ -225,71 +225,33 @@ class CQuickMemoryBase
         iSize = cbTotal;
     }
 
-
-    // Convert UTF8 string to UNICODE string, optimized for speed
-    HRESULT ConvertUtf8_UnicodeNoThrow(const char * utf8str)
-    {
-        bool allAscii;
-        DWORD length;
-
-        HRESULT hr = FString::Utf8_Unicode_Length(utf8str, & allAscii, & length);
-
-        if (SUCCEEDED(hr))
-        {
-            LPWSTR buffer = (LPWSTR) AllocNoThrow((length + 1) * sizeof(WCHAR));
-
-            if (buffer == NULL)
-            {
-                hr = E_OUTOFMEMORY;
-            }
-            else
-            {
-                hr = FString::Utf8_Unicode(utf8str, allAscii, buffer, length);
-            }
-        }
-
-        return hr;
-    }
-
     // Convert UTF8 string to UNICODE string, optimized for speed
     void ConvertUtf8_Unicode(const char * utf8str)
     {
-        bool allAscii;
-        DWORD length;
+        size_t sourceLen = strlen(utf8str);
+        size_t destLen = minipal_get_length_utf8_to_utf16(utf8str, sourceLen, 0);
 
-        HRESULT hr = FString::Utf8_Unicode_Length(utf8str, & allAscii, & length);
+        CHAR16_T* buffer = (CHAR16_T*) AllocThrows((destLen + 1) * sizeof(CHAR16_T));
+        buffer[destLen] = 0;
 
-        if (SUCCEEDED(hr))
+        if (!minipal_convert_utf8_to_utf16(utf8str, sourceLen, buffer, destLen, 0))
         {
-            LPWSTR buffer = (LPWSTR) AllocThrows((length + 1) * sizeof(WCHAR));
-
-            hr = FString::Utf8_Unicode(utf8str, allAscii, buffer, length);
-        }
-
-        if (FAILED(hr))
-        {
-            ThrowHR(hr);
+            ThrowHR(COR_E_INSUFFICIENTMEMORY);
         }
     }
 
     // Convert UNICODE string to UTF8 string, optimized for speed
     void ConvertUnicode_Utf8(const WCHAR * pString)
     {
-        bool allAscii;
-        DWORD length;
-
-        HRESULT hr = FString::Unicode_Utf8_Length(pString, & allAscii, & length);
+        size_t sourceLen = u16_strlen(pString);
+        size_t destLen = minipal_get_length_utf16_to_utf8((const CHAR16_T*)pString, sourceLen, 0);
 
-        if (SUCCEEDED(hr))
-        {
-            LPSTR buffer = (LPSTR) AllocThrows((length + 1) * sizeof(char));
-
-            hr = FString::Unicode_Utf8(pString, allAscii, buffer, length);
-        }
+        LPSTR buffer = (LPSTR) AllocThrows((destLen + 1) * sizeof(char));
+        buffer[destLen] = 0;
 
-        if (FAILED(hr))
+        if (!minipal_convert_utf16_to_utf8((const CHAR16_T*)pString, sourceLen, buffer, destLen, 0))
         {
-            ThrowHR(hr);
+            ThrowHR(COR_E_INSUFFICIENTMEMORY);
```

Comment:
This doesn't seem like the correct error code. We've already allocated the correct amount based on the length computation above, unless that has an error. The failure here shouldn't be insufficient memory or else the issue is in that computation. I would assume a failure here is in conversion not in memory allocation.